keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration, denoting
+  **desired** state of the Pod resources,
+* PodSpec.Container.ResourceBindings (new object) denotes **actual** resources
+  allocated to the Pod by Scheduler,
+* PodStatus.ContainerStatus.ResourcesAllocated (new object) shows the resources
+  held by the Pod and its Containers.
+
+In order to determine the state of a Pod resource update, we add two new
+PodConditions named PodResourcesBound and PodResizeSuccess.
+
+Scheduler sets PodResourcesBound PodCondition, and it can have the following values:
+* Status: false  - Resize request has not been allocated by Scheduler,
+  - Reason: AwaitingPreemption - Lower priority Pods are being pre-empted,
+  - Reason: FailedNodeCapacity - Scheduler determined Node does not have room,
+* Status: true - Scheduler has completed allocating the resize request,
+  - Reason: Accepted - Pod resize approved by Scheduler.
+
+PodResizeSuccess PodCondition can have the following values:
+* Status: true  - Last resize request was successful,
+  - Reason: Success - Pod and its Containers were resized successfully,
+* Status: false - Last resize request failed,
+  - Reason: FailedNodeCapacity - Kubelet determined Node does not have room,
+  - Reason: RejectedResourceQuota - Resize request exceeds ResourceQuota,
+  - Reason: RejectedQoSClassChange - Resize request changes QoS class,
+  - Reason: Retrying - Controller initiated a retry.
+
+#### Container Restart Policy
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizePolicy flag
+for each resource type (CPU, memory) :
+* NoRestart - the default value; resize the Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
```

Comment:
this seems like an argument for the default values for the flag to vary by resource type.

Reply:
yes. I'll update it -  default NoRestart for CPU, and default RestartContainer for memory.